scp/cc/cpio/client_providers/kms_client_provider/src/azure/azure_kms_client_provider.cc

```diff
@@ -0,0 +1,179 @@
+using google::scp::core::HttpHeaders;
+using google::scp::core::HttpMethod;
+using google::scp::core::HttpRequest;
+using google::scp::core::HttpResponse;
+using google::scp::core::Uri;
+using std::make_shared;
+using std::shared_ptr;
+using std::all_of;
+using std::bind;
+using std::cbegin;
+using std::cend;
+using std::placeholders::_1;
+using std::make_pair;
+using std::pair;
+
+namespace google::scp::cpio::client_providers {
+
+static constexpr char kAzureKmsClientProvider[] = "AzureKmsClientProvider";
+
+// We need to take this value from a command line option (It already exists somewhere).
+constexpr char kKMSUnwrapPath[] =
+    "https://127.0.0.1:8000/app/unwrapKey?fmt=tink";
+
+ExecutionResult AzureKmsClientProvider::Init() noexcept {
+  return SuccessExecutionResult();
+}
+
+ExecutionResult AzureKmsClientProvider::Run() noexcept {
+  return SuccessExecutionResult();
+}
+
+ExecutionResult AzureKmsClientProvider::Stop() noexcept {
+  return SuccessExecutionResult();
+}
+
+ExecutionResult AzureKmsClientProvider::Decrypt(
+    core::AsyncContext<DecryptRequest, DecryptResponse>&
+        decrypt_context) noexcept {
+  const auto& ciphertext = decrypt_context.request->ciphertext();
+  if (ciphertext.empty()) {
+    auto execution_result = FailureExecutionResult(
+        SC_AZURE_KMS_CLIENT_PROVIDER_CIPHER_TEXT_NOT_FOUND);
+    SCP_ERROR_CONTEXT(kAzureKmsClientProvider, decrypt_context,
+                      execution_result,
+                      "Failed to get cipher text from decryption request.");
+    decrypt_context.result = execution_result;
+    decrypt_context.Finish();
+    return decrypt_context.result;
+  }
+
+  // Check that there is an ID for the key to decrypt with
+  const auto& key_id = decrypt_context.request->key_resource_name();
+  if (key_id.empty()) {
+    auto execution_result =
+        FailureExecutionResult(SC_AZURE_KMS_CLIENT_PROVIDER_KEY_ID_NOT_FOUND);
+    SCP_ERROR_CONTEXT(kAzureKmsClientProvider, decrypt_context,
+                      execution_result,
+                      "Failed to get Key ID from decryption request.");
+    decrypt_context.result = execution_result;
+    decrypt_context.Finish();
+    return decrypt_context.result;
+  }
+
+  AsyncContext<HttpRequest, HttpResponse> http_context;
+  http_context.request = std::make_shared<HttpRequest>();
+
+  http_context.request->path = std::make_shared<Uri>(kKMSUnwrapPath);
+  http_context.request->method = HttpMethod::POST;
+
+  // Get Attestation Report
+  const auto report = hasSnp() ? fetchSnpAttestation() : fetchFakeSnpAttestation();
+
+  nlohmann::json payload;
+  payload["wrapped"] = ciphertext;
+  payload["kid"] = key_id;
+  payload["attestation"] = report;
+
+  http_context.request->body = core::BytesBuffer(nlohmann::to_string(payload));
+
+  http_context.callback = bind(&AzureKmsClientProvider::OnDecryptCallback,
+                               this, decrypt_context, _1);
+
+  auto execution_result = http_client_->PerformRequest(http_context);
```

Comment:
**[Comment for actual PR]**
We will use `auth_token_provider` for authentication once it's implemented.